### DISCUSSION TITLE
time-zone() doesn't change date-related macros for destination file paths

### DISCUSSION THREAD

**Discussion by zkrx** [syslog-ng/syslog-ng#4073] [Q&A] [answered]
# syslog-ng
## Version of syslog-ng
syslog-ng 3 (3.28.1)
Config version: 3.22
Installer-Version: 3.28.1
Revision: 3.28.1-2+b1
Compile-Date: Nov 19 2020 07:27:16
Module-Directory: /usr/lib/syslog-ng/3.28
Module-Path: /usr/lib/syslog-ng/3.28
Include-Path: /usr/share/syslog-ng/include
Available-Modules: examples,pacctformat,afsocket,azure-auth-header,timestamp,stardate,map-value-pairs,basicfuncs,secure-logging,mod-python.cpython-39-x86_64-linux-gnu,afstomp,afsnmp,afmongodb,dbparser,syslogformat,kafka,add-contextual-data,kvformat,system-source,geoip2-plugin,afsmtp,pseudofile,graphite,csvparser,cryptofuncs,redis,afuser,sdjournal,json-plugin,linux-kmsg-format,disk-buffer,affile,riemann,hook-commands,tfgetent,xml,tags-parser,confgen,appmodel,afsql,http,cef,afamqp,afprog
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: on
Enable-TCP-Wrapper: on
Enable-Linux-Caps: on
Enable-Systemd: on

## Platform
Linux 5.10.0-14-amd64 #1 SMP Debian 5.10.113-1 (2022-04-29) x86_64 GNU/Linux

## Configuration
```
 39 @version: 3.27
 38 @include "scl.conf"
 37 
 36 @define LOCAL_PORT 514
 35 
 34 options {
 33         keep-hostname(yes);
 32         dir-perm(0755);
 31         perm(0755);
 30         time-zone("UTC");
 29 };
 28 
 27 template t_template {
 26         template("${ISODATE} ${HOST} ${MESSAGE}\n");
 25 };
 24 
 23 source s_remote_clients {
 22         syslog(
 21                 ip("0.0.0.0")
 20                 transport("tls")
 19                 tls(
 18                         key-file("/etc/ssl/logserv/keys/logserver.key")
 17                         cert-file("/etc/ssl/logserv/certs/logserver.crt")
 16                         ca-dir("/etc/ssl/logserv/clients")
 15                 )
 14                 port(`LOCAL_PORT`)
 13         );
 12 };
 11 
 10 destination d_tmp {
  9         file(
  8                 "/mnt/syslog-ng/${.SDATA.file@18372.4.hostname}_${.SDATA.file@18372.4.source}_${ISODATE}.tst"
  7                 create_dirs(yes)
  6                 template(t_template)
  5         );
  4 };
  3 
  2 log {
  1         source(s_remote_clients);
  0         destination(d_tmp);
  1 };
```

# Issue
Syslog-ng ignores time-zone information when using date-related macros in the destination file path.

I have multiple clients sending log information to a syslog-ng server. On the server itself, I use a global option `time-zone("UTC")`. I then use a template to rewrite timestamps with the modified timezone inside the log file itself with the `${ISODATE}` macro, which works. However the new time-zone is being ignored when creating the filename with the same `${ISODATE}` macro. This is also the case of the `${HOUR}` macro and I guess other time-related macros as well.

This shouldn't happen according to the documentation:
```
Converting the timezone changes the values of the following date-related macros (macros
MSEC and USEC are not changed):
...
HOUR
...
ISODATE
...
```

## Input and output logs (if possible)
```
Filename: "-rwxr-xr-x 1 root root 1.4K Jul  7 13:09 blah1_bluh1_2022-07-07T13:09:11+02:00.tst"
This file's content: "2022-07-07T11:09:11+00:00 host1 [30:30:23.300,000] <dbg> my message"
```

We can see that the filename hour is set to 13 while the timestamp hour is set to 11. But both use the same `${ISODATE}` macro, so this is inconsistent. The expected behavior should be that both hours are set to 11.

**Comment by bazsi** [comment, marked as answer]
Filenames are expanded using the local time zone of the syslog-ng host.

This time-zone defaults to the local time zone but can be customized with
the local-time-zone() global option, or iirc even at the file() destination.

https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.36/administration-guide/5#TOPIC-1768499


When macro expansions are used in the destination filenames, the local
timezone is used. (Also, if the timestamp of the received message does not
contain the year of the message, syslog-ng OSE uses the local year.)

**Comment by zkrx** [comment]
Ah yes, I missed that option:

```
local-time-zone()

Description: Sets the timezone used when expanding filename and tablename templates.
The timezone can be specified by using the name, for example, time-zone
("Europe/Budapest")), or as the timezone offset in +/-HH:MM format, for example, +01:00).
On Linux and UNIX platforms, the valid timezone names are listed under the
/usr/share/zoneinfo directory.
```

Thank you for pointing this out. I am closing this issue.
